### PR TITLE
Python bindings

### PR DESCRIPTION
**Pull request by AurumTheEnd**

### CODE REVIEW

**Comment by AurumTheEnd** [comment]
@daemontus Any idea why the CI doesn't work? I tried to implement the solution [here](https://pyo3.rs/v0.21.0-beta.0/faq.html?highlight=linker#i-cant-run-cargo-test-or-i-cant-build-in-a-cargo-workspace-im-having-linker-issues-like-symbol-not-found-or-undefined-reference-to-_pyexc_systemerror) but given the lack of a single error message I don't know if this is even the same issue.

**Comment by daemontus** [comment]
The `tarpaulin` thing seems to be a known issue that can be fixed by changing  the engine: https://github.com/xd009642/tarpaulin/issues/1092

Also, it seems that PyO3 0.21 is now stable with a few extra deprecation warnings, so these are now fixed.

**Comment by daemontus** [comment]
Some minor notes:

 - `cd` and `source` don't "really work" inside bash scripts because each command in a script actually runs in a separate shell process. So even though the directory/context technically changes, the change is not reflected in the next command. Using Python virtual environments is therefore a bit of a pain, but it can be done. In most cases, you can just use `.env/bin/python3` instead of `python3` (similar for `pip`) and everything works (i.e. you don't need to "activate" the environment). If you really *need* the environment to be active, you can activate it within the same command (see new `local_dev.sh`).
 
  - I changed `gather_literals_rec` to "return via argument". This is typically much more efficient since a new hash set is not needed for every recursive call.
  
  - I would use `KeyError` instead of `ValueError` when converting `UnknownVariableWhileEvaluating`. Key error is raised when a dictionary does not contain the requested value, which is also happening here. But this is a very minor nitpick :)
  
  - I changed the type of `literal_values` to a `HashMap`. PyO3 can do the same conversion for us automatically.
  
  - I turned the bindings into an optional feature. This means anyone depending on this Rust library won't have to depend on PyO3 by default (and more importantly, compile PyO3 :D). However, bindings should still be covered by CI because we use `--all-features` there.
 
Discussion question:

(1) I'd prefer `evaluate`, `evaluate_checked` and `evaluate_with_default` (with the option to set the default value for unknown variables) in the `Evaluate` trait. Setting unknown variables to `false` by default is a great way to amplify small bugs into hard to detect bugs. This would also affect `evaluate_safe` and `evaluate_nonsafe`. To a Python user, safe and nonsafe will be likely very foreign. We need to have a default `evaluate` method here (that can be also used to implement `__call__`) and throws an exception, and then `evaluate_with_default` in case they want a method that never fails. *I'll prepare a PR for this.*

(2) It would be nice if `UnknownVariableWhileEvaluating` could actually collect *all* invalid variables. This can quickly turn into a whack-a-mole for users if they have to fix errors one-by-one. But this is a very minor issue.

(3) What is `PartialEq` used for on `PythonExpressionError` and why is it implemented like this? (I understand that this is a valid partial equivalence, just not sure why we actually need it).

(4) I now believe that the type that we actually want for the expression is `Expression(Arc<ExpressionData>)` and then have a private `enum ExpressionData { ... }` with all the variants that we have now (using `Expression` instead of `Arc<Expression>` inside). That way, we can make methods like `Expression.rename_literals` that return a new expression when the content actually changed, and simply clone the `Arc` counter if the expression did not change. We could even consider just making `Expression` implement `Copy` and live with that. An obvious issue is that we cannot pattern-match on this. But maybe if the enum was actually public, we could still expose it's reference with something like `expression.data()` and pattern match on that. I am sorry it took me so long to realize this. *I'll make a prototype to see if this is actually as useful as I think it might be.*

**Bottom line: I made a few minor updates but this is "reviewed" and can be merged. I'll be creating issues for the discussion questions to explore later. The coverage drop is ok because we added bindings that aren't covered yet.**
